fundingmanager.go

```diff
@@ -753,6 +800,8 @@ func (f *fundingManager) reservationCoordinator() {
 			}
 		case req := <-f.fundingRequests:
 			f.handleInitFundingMsg(req)
+		case msg := <-f.timeoutMsgs:
```

Comment:
I like this way of handling timeouts much better, but I think it can be simplified significantly.

To me it seems simpler to let each reservation have a `lastUpdated` timestamp. And instead of tracking the state of the funding message and modifying this in every case, I would make it possible to extract the reservation context from the messages in `fundingMsgs`, and each time a message is received, this timestamp can be updated. Now you can add a ticker to the reservation coordinator that at certain intervals loops through all active reservations, and cancels those that haven't been updated in a while. This makes sure the race condition you mentioned cannot occur, as they are never cancelled in parallel with handling messages.

Reply:
I do not think the race condition can occur as it currently stands because, although each reservation's timer is running in parallel, when that timeout occurs it sends a timeout message to the reservation coordinator which handles messages it receives sequentially. Thus, if a timeout occurs after a new message is received for a reservation but before the reservation's timer can be stopped, the timeout will be ignored by `handleTimeoutMsg` because the states will be different. I chose this way of doing timeouts because then any stale reservation will get cleaned up right when it times out, and we do not have to run a "stop the world" zombie sweeper. However, I definitely see your point in that it makes the code more complex.

Anyway, I will write some tests to confirm that there is no race condition and then reduce the code complexity by moving to a zombie sweeper rather than individual timeouts, unless you think that the tradeoff in code complexity is worth avoiding a "stop the world" zombie sweeper.

Any suggestions on how often the zombie sweeper should run?

Reply:
That sounds like a good approach :)

With a 10 minute timeout I would say maybe let it run once a minute?

Reply:
I updated the PR to move to a zombie sweeper implementation, rather than having individual timeouts. I still need to write tests but I wanted to get feedback on whether or not an error message should be sent to the peer when a reservation is timed out and cancelled. I assumed that we would want to send an error message so I included it. However, I left the message generic because of the following [comment in the code](https://github.com/lightningnetwork/lnd/blob/master/fundingmanager.go#L702-L703). Please advise on this. In the meantime, I will add tests.